services/relayer/internal/msgobserver/eth/observer/event_transformer.go

```diff
@@ -84,9 +92,69 @@ func (t *SFCEventTransformer) validate(event *functioncall.SfcCrossCall) error {
 
 // getIDForEvent generates a deterministic ID for an event of the format {network_id}/{contract_address}/{block_number}/{tx_index}/{log_index}
 func (t *SFCEventTransformer) getIDForEvent(event types.Log) string {
-	return fmt.Sprintf("%s/%s/%d/%d/%d", t.Source, t.SourceAddr, event.BlockNumber, event.TxIndex, event.Index)
+	return fmt.Sprintf("%s-%s-%d-%d-%d", t.Source, t.SourceAddr, event.BlockNumber, event.TxIndex, event.Index)
 }
 
 func NewSFCEventTransformer(sourceNetwork string, sourceAddr string) *SFCEventTransformer {
 	return &SFCEventTransformer{sourceNetwork, sourceAddr}
 }
+
+type GPACTEventTransformer struct {
+	Source     string
+	SourceAddr string
+}
+
+func (t *GPACTEventTransformer) ToMessage(event interface{}) (*v1.Message, error) {
+	var raw types.Log
+	startEvent, ok := event.(*functioncall.GpactStart)
+	var funcSig []byte
+	if ok {
+		raw = startEvent.Raw
+		funcSig = startFuncSig[:]
+	} else {
+		segmentEvent, ok := event.(*functioncall.GpactSegment)
+		if ok {
+			raw = segmentEvent.Raw
+			funcSig = segmentFuncSig[:]
+		} else {
+			rootEvent, ok := event.(*functioncall.GpactRoot)
+			if ok {
+				raw = rootEvent.Raw
+				funcSig = rootFuncSig[:]
+			} else {
+				return nil, fmt.Errorf("event not supported")
+			}
+		}
```

Comment:
Should we have separate transformers for start and segment events instead? I think it might be cleaner and probably more efficient. The different transformers can be attached to the specific watchers for those event types.

Reply:
Yes, will do in next PR.